p2p/src/peer_manager/mod.rs

```diff
@@ -1090,29 +1145,107 @@ where
+            let addr_group = AddressGroup::from_peer_address(&addr.as_peer_address());
+
+            match role {
+                PeerRole::Inbound => {}
+                PeerRole::OutboundManual => {
+                    // TODO: should we include manual peer connection in cur_outbound_conn_addr_groups,
+                    // in order to avoid opening new automatic connections to their address groups?
+                    // (Bitcoin does it).
+                    // See the TODO section of https://github.com/mintlayer/mintlayer-core/issues/832
+                    // Note that this change will require adjusting expected connections numbers
+                    // in the "discovered_node" tests.
+                }
+                PeerRole::OutboundFullRelay => {
+                    cur_outbound_full_relay_conn_count += 1;
+                    cur_outbound_conn_addr_groups.insert(addr_group);
+                }
+                PeerRole::OutboundBlockRelay => {
+                    cur_outbound_block_relay_conn_count += 1;
+                    cur_outbound_conn_addr_groups.insert(addr_group);
+                }
+            }
         }
 
-        self.try_evict_block_relay_peer();
+        let needed_outbound_full_relay_conn_count = {
+            let extra_conn_count = if self.tip_is_stale() {
+                *self.p2p_config.peer_manager_config.outbound_full_relay_extra_count
+            } else {
+                0
+            };
 
-        if let Some(o) = self.observer.as_mut() {
-            o.on_heartbeat();
+            (*self.p2p_config.peer_manager_config.outbound_full_relay_count + extra_conn_count)
+                .saturating_sub(cur_outbound_full_relay_conn_count)
+        };
+
+        let new_full_relay_conn_addresses = self.peerdb.select_new_non_reserved_outbound_addresses(
+            &cur_outbound_conn_addr_groups,
+            needed_outbound_full_relay_conn_count,
+        );
+
+        // TODO: in bitcoin they also try to create an extra outbound full relay connection
+        // to an address in a reachable network in which there are no outbound full relay or
+        // manual connections (see CConnman::MaybePickPreferredNetwork for reference).
+        // See the TODO section of https://github.com/mintlayer/mintlayer-core/issues/832
+
+        for address in new_full_relay_conn_addresses.into_iter() {
+            let addr_group = AddressGroup::from_peer_address(&address.as_peer_address());
+            cur_outbound_conn_addr_groups.insert(addr_group);
+
+            self.connect(
+                address,
+                OutboundConnectType::Automatic {
+                    block_relay_only: false,
+                },
+            );
+        }
+
+        let needed_outbound_block_relay_conn_count =
+            (*self.p2p_config.peer_manager_config.outbound_block_relay_count
+                + *self.p2p_config.peer_manager_config.outbound_block_relay_extra_count)
```

Comment:
Unless you insist, I'd rather keep it as is, because I'm not sure what would be a good name for this thing, provided that we also have the corresponding "full_relay_count/full_relay_extra_count" fields, for which such a sum doesn't always make sense.

Reply:
no problem